src/backend/distributed/planner/local_plan_cache.c

```diff
@@ -99,6 +96,66 @@ CacheLocalPlanForShardQuery(Task *task, DistributedPlan *originalDistributedPlan
 }
 
 
+/*
+ * GetLocalShardQuery is a helper function which generates the local shard
+ * query based on the jobQuery.
+ *
+ * It is not guaranteed to have consistent attribute numbers on the shards
+ * and on the shell (e.g., distributed/reference tables) due to DROP COLUMN
+ * commands. To avoid any edge cases due to such discrepancies, we first
+ * deparse the jobQuery with the tables replaced to shards, and parse the
+ * query string back. This is normally a very expensive operation, however
+ * we only do it once per cached local plan, which is acceptable.
+ */
+static Query *
+GetLocalShardQuery(Query *jobQuery, Task *task, ParamListInfo orig_paramListInfo)
+{
+	StringInfo localShardSqlQuery = makeStringInfo();
+	if (jobQuery->commandType == CMD_INSERT)
+	{
+		/*
+		 * For INSERT queries we cannot use pg_get_query_def. Mainly because we
+		 * cannot run UpdateRelationToShardNames on an INSERT query. This is
+		 * because the PG deparsing logic fails when trying to insert into a
+		 * RTE_FUNCTION (which is what will happen if you call
+		 * UpdateRelationToShardNames).
+		 */
+		deparse_shard_query(jobQuery, task->anchorDistributedTableId, task->anchorShardId,
+							localShardSqlQuery);
```

Comment:
It seems like this logic is not robust to INSERT..SELECT. Should we have some Assert? Alternatively, what if we call UpdateRelationToShardNames here as well? That seems to be how we handle INSERT..SELECT via PlanRouterQuery.

Reply:
`UpdateRelationToShardNames((Node *) jobQuery, task->relationShardList);` followed by `TaskQueryString(task);` seems to be more generic, switching to that makes more sense

Reply:
Also, in `DeparseTaskQuery` there is a comment saying:
```		 * For INSERT queries we cannot use pg_get_query_def. Mainly because we
		 * cannot run UpdateRelationToShardNames on an INSERT query. This is
		 * because the PG deparsing logic fails when trying to insert into a
		 * RTE_FUNCTION (which is what will happen if you call
		 * UpdateRelationToShardNames).
		 */
```

However, pg deparsing logic doesn't seem to fail whenn we insert RTE_FUNCTION for INSERT. Is tthat something we improved in the past? Or the comment is not accurate?

Reply:
well, it seems this is accurate, we cannot call `UpdateRelationToShardNames` for plain INSERTs. so, switched back to old implementation, and added an assert for now. I'll look more and update here